crates/block-producer/src/psc.rs

```diff
@@ -736,6 +766,79 @@ async fn median_gte(rpc_client: &RPCClient, timestamp_millis: u64) -> Result<()>
     Ok(())
 }
 
+/// Submit next pending l1 upgrade
+async fn submit_pending_l1_upgrade(ctx: &PSCContext) -> Result<()> {
+    let snap = ctx.store.get_snapshot();
+    let block_number = snap
+        .get_last_confirmed_block_number_hash()
+        .expect("last confirmed")
+        .number()
+        .unpack()
+        + 1;
+    drop(snap);
+
+    // find next l1 upgrade by block number
+    if let Some(l1_upgrade) = ctx
+        .rollup_context()
+        .fork_config
+        .pending_l1_upgrades
+        .iter()
+        .find(|l1_upgrade| l1_upgrade.height == block_number)
+    {
+        let tx: Transaction = l1_upgrade.signed_transaction.clone().into();
+        // update local cell and status
+        ctx.local_cells_manager
+            .lock()
+            .await
+            .apply_tx(&tx.as_reader());
+
+        log::info!(
+            "sending l1 upgrade transaction 0x{}",
+            hex::encode(tx.calc_tx_hash().as_slice())
+        );
+
+        // We can't recover from errors when sending l1 upgrade tx, so we just throw it
+        send_transaction_or_check_inputs(&ctx.rpc_client, &tx).await?;
+        log::info!("tx sent");
+
+        // Refresh block producer's contracts deps
+        ctx.block_producer
+            .contracts_dep_manager()
+            .refresh(Some(tx))
+            .await?;
```

Comment:
Do we really need to add the transaction to local cells manager if we are directly getting deps from it? The upgrade tx shouldn't directly involve any cells used/created by submission txs.

Reply:
(Assuming that the upgrade tx doesn't use the same wallet/lock as the block producer to pay fee.)

Reply:
Removed the local cell manager injection